### PR TITLE
[8.x] Allow model resources to define their own response class

### PR DESCRIPTION
**Pull request by huskicsuad**
### Motivation:
We have strict API standards that we need to follow in our responses, for pagination; we need to return a single `cursor` value. I would like to create a base pagination resource response class that our application can use for all resources.

### Benefits:
This will give folks more flexibility while maintaining BC

Resource response to be resolved by the container, allowing folks to DI any other classes into their response class -- for example a custom cursor.

### CODE REVIEW

**Comment by deleugpn** [comment]
Have you considered using the `Responsable` interface instead of API Resource?

**Comment by huskicsuad** [comment]
@deleugpn Yes, but i would like to utilize the helpers that come with API resources, `mergeWhen`, `mergeWhenLoaded`, etc.. I also want us to continue to follow the framework way, easier for new folks to understand since laravel is well documented.

**Comment by morloderex** [comment]
@huskicsuad You might not even need this sense the current JsonResource and `ResourceCollection` classes both contains an empty `withResponse` method. You might be able to utilize this method to define what ever logic you desire. 

You have access to the original content within the property `original` on the response object.

Not sure tho just a thought.

**Comment by huskicsuad** [comment]
@morloderex -- can yo share more details on that? I tried messing around with `withResponse` but it ended up being very hacky

File: src/Illuminate/Http/Resources/Json/JsonResource.php [hunk 1]
```diff
@@ -218,7 +226,34 @@ public function response($request = null)
...
+     * @return ResourceResponse
+     */
+    protected function resolveResourceResponse()
+    {
+        if ($this->resourceResponse) {
+            $resourceResponse = $this->resourceResponse;
+        } else {
+            $resourceResponse = $this->isResourcePaginated()
+                ? PaginatedResourceResponse::class
+                : ResourceResponse::class;
```
Comment (taylorotwell) [inline src/Illuminate/Http/Resources/Json/JsonResource.php:243]: Why did you change the behavior of the base JsonResource class to do this pagination stuff? It was not like that before. Please *only* make the changes necessary for the feature you want.
